### ISSUE TITLE
what are those "void**" for argument and return register of a function? 

### ISSUE THREAD

**Issue by hlide** [yegord/snowman#56]
This simple *strlen* implementation for Allegrex:
```
89014c4:	move       $v1, $a0
89014c8:	lb         $v0, 0($v1)
89014cc:	bnez       $v0, 0x089014C8
89014d0:	addiu      $v1, $v1, 1
89014d4:	nor        $v0, $zr, $a0
89014d8:	jr         $ra
89014dc:	addu       $v0, $v1, $v0
```
where only registers *a0*, *v1* are modified *v0*, gave me that strange c++ output:
```c++
void** strlen(unsigned char* a0, void** a1, void** a2, void** a3, void** t0, void** t1, void* t2) {
    unsigned char* v1_8;

    v1_8 = a0;
    while (*v1_8) {
        ++v1_8;
    }
    return (uint32_t)(v1_8 + 1) + ~(uint32_t)a0;
}
```
(Allegrex can have up to 8 argument in register (*a0-a3, t0-t3*).

**Comment by nihilus** [comment]
yeah, I noticed these quirky function arguments as well. 
MIPS ELF files are kinda obscure and we need to improve the ELF parser. MIPS PE32 files does not show this behaviour.

**Comment by yegord** [comment]
If you are decompiling the function alone, this is strange. You can give me the input file and the git commit, I can have a look.

If you are decompiling the whole program, these are just wrong guesses of the heuristic guessing the call arguments at the call site.

**Comment by hlide** [comment]
No, I'm not decompiling the function alone - it is all the program which is decompiled.

**Comment by nihilus** [comment]
The HPPA and MIPS ELF ABI is kinda obscure when it comes to loading external functions but for an internal function it should work.

**Comment by JustEnuff2BDangerous** [comment]
Not sure if this helps or not but...

Looking at online references (LLVM source etc.) Allegrex is a mash up of different MIPS architectural feature which looking briefly at nihilus' fork he's covered at the instruction level and there isn't anything special I can immediately see in the ABI thats needed to support them and would cause this issue. 

The one thing I did spot is that Allegrex doesn't use the most common ABI's for MIPS (O32/N32) it actually uses the embedded variant (EABI). This should be easily confirmed by checking the e_flags in the elf program header (see MIPS extensions to ELF [here] (https://dmz-portal.mips.com/wiki/MIPS_ELF_header_definitions) )

I only found one references to MIPS EABI variant (see [here] (http://www.cygwin.com/ml/binutils/2003-06/msg00436.html) ) and at first glance its not radically different to the O32 default. If you decompile the function in isolation and it produces the correct output then its likely that the source of the problem in whole program decompilation attempting to discover the functional signature of the target function from the calling function rather then the target. The hints are there in the output... 

* temp registers (t<n>) appear in the func sig. but ABI never uses them for parameter passing
** However they maybe stacked before calling another func in order to preserve func state.

External (DLL) functions shouldn't require any different handling as they'll probably branch through a PLT stub (you can do GOT direct branching, but its rarely seen), this should be effectively transparent from the ABI point of view unless you have ABI veneering (common in ARM) which manage differences between ABI's. However since Allegrex is a "closed" platform I wouldn't expect that to be the case.

**Comment by hlide** [comment]
I used to code some PSP programs in the glory times of PSP (you know when Customized Firmware ruled...) and one thing I noticed is registers are using O32 naming but are used as N32 registers. That is, t0-t3 are indeed used as a4-a7 (so up to 8 integer arguments for a PSP functions, the rest on stack).
